### PR TITLE
Add support for Almquist shell in masking

### PR DESCRIPTION
**Pull request by Wadeck**
Almquist shell is the default one for Alpine. The escape schema is a bit special compared to the other.

```
pwd="pass'word"
set -x
echo $pwd
```
outputs
`+ echo 'pass'"'"'word'`

You can reproduce those commands on `jenkinsci/jnlp-slave:alpine` default shell.

Unit tests added.

For the review information: `'` is escaped as `'"'"'`, `''` as `'"''"'`, `'''` as `'"'''"'`

@jvz

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        return passwords;
+    }
+
+    @ClassRule public static JenkinsRule j = new JenkinsRule();
+
+    private WorkflowJob project;
+    private String credentialsId;
+
+    @BeforeClass
+    public static void assumeBash() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]:
```suggestion
    public static void assumeAsh() {
```

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        CredentialsTestUtil.setStringCredentials(j.jenkins, credentialsId, credentials);
+        WorkflowRun run = runProject();
+
+        j.assertLogContains("begin0 **** end0", run);
+        j.assertLogContains("begin2 **** end2", run);
+        j.assertLogNotContains(credentials, run);
+    }
+
+    private WorkflowRun runProject() throws Exception {
+        return j.assertBuildStatusSuccess(project.scheduleBuild2(0));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]:
```suggestion
        return j.buildAndAssertSuccess(project);
```

which is so short it should just be inlined anyway.

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    private WorkflowJob project;
+    private String credentialsId;
+
+    @BeforeClass
+    public static void assumeBash() {
+        // ash = Almquist shell, default one used in Alpine
+        assumeThat("ash", is(executable()));
+        // due to https://github.com/jenkinsci/durable-task-plugin/blob/e75123eda986f20a390d92cc892c3d206e60aefb/src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java#L149
+        // on Windows
+        assumeThat("nohup", is(executable()));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: Huh? Why is this here? The link does not explain anything to me.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactory.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import hudson.Extension;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * Almquist shell, aka ash, is the default shell in Alpine distribution
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactory.java]:
```suggestion
 * Almquist shell, aka ash, is the default shell in Alpine distribution.
```

(`javadoc` cares about punctuation.)

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.experimental.theories.DataPoint;
+import org.junit.experimental.theories.DataPoints;
+import org.junit.experimental.theories.Theories;
+import org.junit.experimental.theories.Theory;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: Is there a particular reason you are using this “experimental” facility? AFAICT the test could be written just as well using the better-known `Parameters`.

Comment (Wadeck) [reply src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: Matt provided that approach in the previous test for Bash/Bourne shell, so I re-used them. I was a bit disappointed to see that the theories are failing as soon as one dataPoint failed compared to parameter.

This could be changed in a future version without problem (along with the other usage of that approach).

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java (hunk 3 above)
Comment (Wadeck) [reply src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: The link points to a code adding the `nohup` argument in the command. On Windows that command is not necessary present (at least not in my environments) and thus just checking for the shell is not sufficient when used in `durable-task`.

Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: I would not expect `ash` to be present on Windows either.

Comment (Wadeck) [reply src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: Good point 😕

Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/credentialsbinding/masking/AlmquistShellSecretPatternFactoryTest.java]: Anyway, it is OK, just a bit confusing.

**Comment by jglick** [comment]
[INFRA-2075](https://issues.jenkins-ci.org/browse/INFRA-2075)

**Comment by Wadeck** [comment]
Manually tested some different scenarios, you can merge it :)
